docs/src/adr/008adr-exceptions.md

```diff
@@ -0,0 +1,122 @@
+
+### 1.1. Parser-level
+Since we depend on Sany for parsing, Apalache rejects any syntax which Sany cannot parse.
+If Sany produces an exception Apalache catches it and re-throws a class extending
+
+```
+ParserException
+```
+
+The exception should report the following details:
+
+  * Location of at least one parser issue
+
+### 1.2. Apalache-specific input exceptions
+This category of exceptions deals with problems triggered by incorrect or incomplete information regarding Apalache inputs. Examples include:
+
+  * Malformed config files
+  * Incorrect or missing `OVERRIDE_`
+  * Incorrect or missing `UNROLL_`
+  * Problems with `--init/next/cinit/...`
+
+Exceptions thrown in response to these issues should extend 
+
+```
+ApalacheInputException
+```
+
+The exceptions should report the following details:
+
+  * If an input is missing: the name of the expected input 
+  * If an input is incorrect: the way in which it is incorrect
+
+### 1.3. Type-related exceptions
+This category of exceptions deals with problems arising from the type-system used by Apalache. Examples include:
+
+  * Missing or incorrect type annotations
+  * Incompatibility of argument types and operator types
+  
+Exceptions thrown in response to these issues should extend 
+
+```
+TypeingException
+```
+
+The exceptions should report the following details:
+
+  * If an annotation is missing: the declaration with the missing annotation and its location  
+  * If the types of the arguments at a built-in operator application site are incompatible with the operator type: Both the computed and expected types, and the location of the application site 
+  * If the types of the arguments at a user-defined operator application site are incompatible with the operator annotation: Both the computed and expected types, the operator declaration, and the location of the application site
+
+### 1.4. Static-analysis exceptions
+This category of exceptions deals with problems arising from the various static analysis passes performed by Apalache. Examples include:
+
+  * Missing or incorrect variable assignments
+  * Other analyses we might run in the future
+  
+Exceptions thrown in response to these issues should extend 
+
+```
+StaticAnalysisException
+```
+
+The exceptions should report the following details:
+
+  * The location in the specification where the analysis failed and the expected result
+
+### 1.5. Unsupported language exceptions
+This category of exceptions deals with user input, which falls outside of the TLA+ fragment supported by Apalache. Examples include:
+
+  * Unbounded quantification
+  * (Unbounded) `CHOOSE`
+  * `SelectSeq`
+  * Fragments of community modules
+
+Exceptions thrown in response to these issues should extend 
+
+```
+UnsupportedFeatureException
+```
+
+The exceptions should report the following details:
+
+  * The location of the unsupported expression(s)
+
```

Comment:
Seems this doesn't include user errors at the CLI level.

Reply:
Those fall under 1.2, no? `Problems with `--init/next/cinit/...`

Reply:
Ah, I missed that. I guess it does fall under inputs.